### ISSUE TITLE
Proposal for the initial API

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire-ml#3]
This library will sit on top of arrayfire. The default implementation will be C++, however C wrappers will be added on top so that the library can be used from other languages.

The optimization function and distance metrics need to be abstract out so the user has a choice of switching them out whenever possible.
## Base classes
- [ ] [Classifier](https://github.com/arrayfire/arrayfire_ml/issues/16) (Model fit ?)
- [ ] [Neural Networks](https://github.com/arrayfire/arrayfire_ml/issues/17)
### Classifiers
- [ ] [Perceptron](https://github.com/arrayfire/arrayfire_ml/issues/4)
- [ ] [Naive Bayes classifier](https://github.com/arrayfire/arrayfire_ml/issues/5)
- [ ] [Support Vector Machines](https://github.com/arrayfire/arrayfire_ml/issues/6) (SVM)
- [ ] [Logistic Regression](https://github.com/arrayfire/arrayfire_ml/issues/7)
- [ ] [Softmax Regression](https://github.com/arrayfire/arrayfire_ml/issues/7)
- [ ] [Multi Layer Perceptron](https://github.com/arrayfire/arrayfire_ml/issues/8)
- [ ] [Convolutional Neural Networks](https://github.com/arrayfire/arrayfire_ml/issues/1) (CNN)
- [ ] [Deep Belief Networks](https://github.com/arrayfire/arrayfire_ml/issues/10)
- [ ] [Nearest Neighbour Search](https://github.com/arrayfire/arrayfire_ml/issues/9) (kNN)
### Dimensionality Reduction
- [ ] [Restricted Boltzman Machines](https://github.com/arrayfire/arrayfire_ml/issues/11) (RBM)
- [ ] [Autoencoders](https://github.com/arrayfire/arrayfire_ml/issues/11)
- [ ] [Principal Component Analysis](https://github.com/arrayfire/arrayfire_ml/issues/18) (PCA)
- [ ] [Linear Discriminant Analysis](https://github.com/arrayfire/arrayfire_ml/issues/19) (LDA)
- [ ] [Locality Sensitive Hashing](https://github.com/arrayfire/arrayfire_ml/issues/12) (LSH)
### Regression Analysis
- [ ] [Linear Regression](https://github.com/arrayfire/arrayfire_ml/issues/13)
- [ ] [Genaralized Linear Regression](https://github.com/arrayfire/arrayfire_ml/issues/14) (GLM)
### Clustering
- [ ] [k-Means](https://github.com/arrayfire/arrayfire_ml/issues/15)
### Optimization functions

Needs research

**Comment by pavanky** [comment]
@umar456 @pentschev @jramapuram Any suggestions about what can go in the first go would be greatly appreciated.

**Comment by pavanky** [comment]
In particular, additions to Regression and clustering would be nice.

**Comment by jramapuram** [comment]
I think it makes sense to start with the basics, go with:
- Linear Regression
- Logistic Regression (natural extension due to just a kernel swap)
- k-Means 
- Standard Multi Layer Perceptron.

And then take it from there! :)

Personally I would like to maybe do some LSTM's & other RNN type models.
However, for these we will need an optimizer. SGD & [ADAM](http://arxiv.org/abs/1412.6980) should be easy enough to implement

**Comment by jramapuram** [comment]
Btw for the suite of neural models it would be good to abstract a standard 'Layer' class. This will make the addition of conv layers/etc pretty easy.

**Comment by pavanky** [comment]
@jramapuram I was just running through the library design via github issues. I came up with this: https://github.com/arrayfire/arrayfire_ml/issues/17 for the standard Layer class. It might change once the work actually begins.

**Comment by pavanky** [comment]
I will look into the optimization functions and update the list. Can you create an issue for base RNN class and / or LSTM and discuss what additional work should be done ?

**Comment by jramapuram** [comment]
Sure thing. For optimization I suggest paralleling something like [keras's solution](https://github.com/fchollet/keras/blob/master/keras/optimizers.py) minus maybe the automatic differentiation that theano offers (thats pretty tough to implement I'd imagine).

**Comment by futurely** [comment]
@zxie has implemented a very elegant [neural network library](https://github.com/zxie/nn) which maybe was based on the [general flow graph](http://cs224d.stanford.edu/lectures/CS224d-Lecture5.pdf) described in the Stanford course [Deep Learning for Natural Language Processing](http://cs224d.stanford.edu/syllabus.html).

A [graph](https://github.com/zxie/nn/blob/master/nets/graph.py) based neural network model can be built with [Nodes](https://github.com/zxie/nn/blob/master/nets/nodes.py) which do the math with [Ops](https://github.com/zxie/nn/blob/master/gpu/ops.py). Currently [only one model](https://github.com/zxie/nn/blob/master/nets/nplm_graph.py) was implemented with nodes. The library also implements some NNet models without using the nodes.

Caffe is also [trying to manage the relationship of layers](https://github.com/BVLC/caffe/pull/2073) in a computation graph.

**Comment by unbornchikken** [comment]
I have some experience with recurrent learning models, like RTLR and LSTM. Those coluld be used for tasks like speech recognition and realtime time series predictions, and so. Do you have plans for supporting such functionality aswell?

**Comment by jramapuram** [comment]
@unbornchikken : https://github.com/arrayfire/arrayfire_ml/issues/20

**Comment by unbornchikken** [comment]
Yeah, but that doesn't get icluded in this list above. Yet?

**Comment by UniqueFool** [comment]
> Caffe is also trying to manage the relationship of layers in a computation graph.

FWIW, there is an increasing number of ML algorithms that are doing tree/graph processing to some extent, so that may be another useful thing to have in arrayfire-ml: graph processing utilities for some of the more common algorithms (e.g. [breadth-first search](https://github.com/otrebor/JGA/blob/master/kernels/kernelsBFS.cl)  implemented in OpenCL)

**Comment by pavanky** [comment]
Closing this because it is no longer relevant.

**Comment by jasonhargrove** [comment]
Why is it no longer relevant?

**Comment by pavanky** [comment]
@jasonhargrove Using autograd for implementing neural networks. This involved changing the API. If you have any suggestions / questions, please follow up over here:
https://github.com/arrayfire/arrayfire-ml/issues/17

**Comment by tcandzq** [comment]
Hi，Could I join this project and develop some function in arrayfire-ml。Because write ml algorithm in C++ and GPU would interesting and challenging! Thank you!

**Comment by dimka8** [comment]
Good day . ubuntu 20.04 error will occur
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Internal error:998):
In function compileModule
In file src/backend/cuda/compile_module.cpp:279
NVRTC Error(7): NVRTC_ERROR_BUILTIN_OPERATION_FAILURE
Log: 
nvrtc: error: failed to open libnvrtc-builtins.so.11.4.
  Make sure that libnvrtc-builtins.so.11.4 is installed correctly.

 0# 0x00007F5B6FA6E9B0 in /opt/arrayfire/lib64/libafcuda.so.3
 1# 0x00007F5B70CFCBA2 in /opt/arrayfire/lib64/libafcuda.so.3
 2# 0x00007F5B6FEC806C in /opt/arrayfire/lib64/libafcuda.so.3
 3# 0x00007F5B6FECC191 in /opt/arrayfire/lib64/libafcuda.so.3
 4# 0x00007F5B6FECD017 in /opt/arrayfire/lib64/libafcuda.so.3
 5# 0x00007F5B6FA09269 in /opt/arrayfire/lib64/libafcuda.so.3
 6# af_eval in /opt/arrayfire/lib64/libafcuda.so.3
 7# af::array::eval() const in /opt/arrayfire/lib64/libafcuda.so.3

all ArrayFire examples work . cuda-11.4 installed

**Comment by dimka8** [comment]
ArrayFire v3.9.0 (CUDA, 64-bit Linux, build 1d03d07)
Platform: CUDA Runtime 11.4, Driver: 510.47.03
[0] NVIDIA GeForce GT 1030, 2001 MB, CUDA Compute 6.1
ArrayFire Exception (Internal error:998):
In function compileModule
In file src/backend/cuda/compile_module.cpp:279
NVRTC Error(7): NVRTC_ERROR_BUILTIN_OPERATION_FAILURE
Log: 
nvrtc: error: failed to open libnvrtc-builtins.so.11.4.
  Make sure that libnvrtc-builtins.so.11.4 is installed correctly.

 0# 0x00007F411B8659B0 in /opt/arrayfire/lib64/libafcuda.so.3
 1# 0x00007F411CAF3BA2 in /opt/arrayfire/lib64/libafcuda.so.3
 2# 0x00007F411BCBF06C in /opt/arrayfire/lib64/libafcuda.so.3
 3# 0x00007F411BCCA511 in /opt/arrayfire/lib64/libafcuda.so.3
 4# 0x00007F411BCCB397 in /opt/arrayfire/lib64/libafcuda.so.3
 5# 0x00007F411B8065B9 in /opt/arrayfire/lib64/libafcuda.so.3
 6# 0x00007F411C0EDFD1 in /opt/arrayfire/lib64/libafcuda.so.3
 7# af_where in /opt/arrayfire/lib64/libafcuda.so.3
 8# af::where(af::array const&) in /opt/arrayfire/lib64/libafcuda.so.3
 9# 0x000055C9C7A1CD5C in ./perceptron_cuda
10# perceptron_demo(bool, int) in ./perceptron_cuda
11# main in ./perceptron_cuda
12# __libc_start_main in /lib/x86_64-li
